src/mcts/node.cc

```diff
@@ -555,11 +555,48 @@ V6TrainingData Node::GetV6TrainingData(
   std::fill(std::begin(result.probabilities), std::end(result.probabilities),
             -1);
   // Set moves probabilities according to their relative amount of visits.
+  // Compute Kullback-Leibler divergence in nats (between policy and visits).
+  float kld_sum = 0;
+  float max_p = -std::numeric_limits<float>::infinity();
+  std::vector<float> intermediate;
+  if (nneval) {
+    int last_idx = 0;
+    for (const auto& child : Edges()) {
+      auto nn_idx = child.edge()->GetMove().as_nn_index(transform);
+      float p = 0;
+      for (int i = 0; i < nneval->p.size(); i++) {
+        // Optimization: usually moves are stored in the same order as queried.
+        const auto& move = nneval->p[last_idx++];
+        if (last_idx == nneval->p.size()) last_idx = 0;
+        if (move.first == nn_idx) {
+          p = move.second;
+          break;
+        }
+      }
+      intermediate.emplace_back(p);
+      max_p = std::max(max_p, p);
+    }
+  }
+  float total = 0.0;
+  auto it = intermediate.begin();
   for (const auto& child : Edges()) {
-    result.probabilities[child.edge()->GetMove().as_nn_index(transform)] =
-        total_n > 0 ? child.GetN() / static_cast<float>(total_n) : 1;
+    auto nn_idx = child.edge()->GetMove().as_nn_index(transform);
+    float fracv = total_n > 0 ? child.GetN() / static_cast<float>(total_n) : 1;
+    if (fracv > 0 && nneval) {
+      float P = std::exp((*it - max_p) / temp);
+      kld_sum += fracv * std::log(fracv / P);
+      total += P;
+      it++;
```

Comment:
'it' should be advance unconditionally, not only if fracv is positive.

Reply:
Argh!